server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -262,6 +269,39 @@ protected void doClose() {
         }
     }
 
+    // Inspects all cluster state elements that contain a hint about what the current repository generation is and updates
+    // #latestKnownRepoGen if a newer than currently known generation is found
+    @Override
+    public void updateState(ClusterState state) {
+        if (readOnly) {
+            // No need to waste cycles, no operations can run against a read-only repository
+            return;
+        }
+        long bestGenerationFromCS = RepositoryData.EMPTY_REPO_GEN;
+        final SnapshotsInProgress snapshotsInProgress = state.custom(SnapshotsInProgress.TYPE);
+        if (snapshotsInProgress != null) {
+            bestGenerationFromCS = bestGeneration(snapshotsInProgress.entries());
+        }
+        final SnapshotDeletionsInProgress deletionsInProgress = state.custom(SnapshotDeletionsInProgress.TYPE);
+        if (bestGenerationFromCS == RepositoryData.EMPTY_REPO_GEN && deletionsInProgress != null) {
+            bestGenerationFromCS = bestGeneration(deletionsInProgress.getEntries());
+        }
+        final RepositoryCleanupInProgress cleanupInProgress = state.custom(RepositoryCleanupInProgress.TYPE);
+        if (bestGenerationFromCS == RepositoryData.EMPTY_REPO_GEN && cleanupInProgress != null) {
+            bestGenerationFromCS = bestGeneration(cleanupInProgress.entries());
+        }
+
+        final long finalBestGen = bestGenerationFromCS;
+        latestKnownRepoGen.updateAndGet(known -> Math.max(known, finalBestGen));
+    }
+
+    private long bestGeneration(Collection<? extends RepositoryOperation> operations) {
+        final String repoName = metadata.name();
+        assert operations.size() <= 1 : "Assumed one or no operations but received " + operations;
+        return operations.stream().filter(e -> e.repository().equals(repoName)).map(RepositoryOperation::repositoryStateId)
+            .findFirst().orElse(RepositoryData.EMPTY_REPO_GEN);
```

Comment:
`mapToLong(RepositoryOperation::repositoryStateId).max().orElse(RepositoryData.EMPTY_REPO_GEN);`?

Reply:
Jup that looks nicer :) Fixed now